larray/core/array.py

```diff
@@ -6549,6 +6545,173 @@ def split_axes(self, axes=None, sep='_', names=None, regex=None):
     split_axis = renamed_to(split_axes, 'split_axis')
 
 
+def parse(s):
+    """
+    Used to parse the "folded" axis ticks (usually periods).
+    """
+    # parameters can be strings or numbers
+    if isinstance(s, basestring):
+        s = s.strip()
+        low = s.lower()
+        if low == 'true':
+            return True
+        elif low == 'false':
+            return False
+        elif s.isdigit():
+            return int(s)
+        else:
+            try:
+                return float(s)
+            except ValueError:
+                return s
+    else:
+        return s
+
+
+def df_labels(df, sort=True):
+    """
+    Returns unique labels for each dimension.
+    """
+    idx = df.index
+    if isinstance(idx, pd.core.index.MultiIndex):
+        if sort:
+            return list(idx.levels)
+        else:
+            return [list(unique(idx.get_level_values(l))) for l in idx.names]
+    else:
+        assert isinstance(idx, pd.core.index.Index)
+        # use .values if needed
+        return [idx]
+
+
+def cartesian_product_df(df, sort_rows=False, sort_columns=False, **kwargs):
+    labels = df_labels(df, sort=sort_rows)
+    if sort_rows:
+        new_index = pd.MultiIndex.from_product(labels)
+    else:
+        new_index = pd.MultiIndex.from_tuples(list(product(*labels)))
+    columns = sorted(df.columns) if sort_columns else list(df.columns)
+    # the prodlen test is meant to avoid the more expensive array_equal test
+    prodlen = np.prod([len(axis_labels) for axis_labels in labels])
+    if prodlen == len(df) and columns == list(df.columns) and np.array_equal(df.index.values, new_index.values):
+        return df, labels
+    return df.reindex(new_index, columns, **kwargs), labels
+
+
+def from_series(s, sort_rows=False):
```

Comment:
Do we really need to have from_series, from_frame et al in here? They belong more to the io package.

Reply:
IMO, "I/O" means read and write from files but such definition is maybe too restrictive...

Reply:
i/o means... input and output. IMO any conversion from/to an external format (even living in memory) falls under i/o. Otherwise, you get some weird inconsitencies: to_excel('file.xlsx') is i/o under your definition but to_excel() (live excel instance) is not. read_csv('some_url') is not i/o, etc.

Reply:
make more sense indeed.